prometheus/timer_test.go

```diff
@@ -0,0 +1,41 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus
+
+import (
+	"testing"
+
+	dto "github.com/prometheus/client_model/go"
+)
+
+func TestTimerObserve(t *testing.T) {
+	his := NewHistogram(
+		HistogramOpts{
+			Name: "test_histogram",
+		},
+	)
+
+	stop := NewTimer()
+	stop(his)
```

Comment:
> The advantage of this method is that the metric type and the timer are independent.

I'm not sure that is an advantage, it makes discovery harder.

> As long as a metric implements the Observe(int64) method, it works with the timer.

We also need this on gauges for batch job timing, it's startTimer/setDuration in Java.

> The main disadvantage of your example is that each metric that wants to have this helper has to have the additional methods StartTimer and ObserveDuration defined on it.

A bit of extra code in a client to make usage easier to a very common use case seems like a good tradeoff to me.